### ISSUE TITLE
Asynchronous start of input plugins

### ISSUE THREAD

**Issue by andrewvc** [elastic/logstash#4931]
This line of inquiry was inspired by the question here: https://github.com/logstash-plugins/logstash-input-rabbitmq/issues/58

The following use case does not make much sense, but this is how it works today:

1. Logstash connects to rabbitmq server starts ingesting logs
1. RabbitMQ Goes down, logstash repeatedly retries connecting
1. Connection attempt succeeds, logstash succeeds in processing logs.

The above is great, however, if in step 2. Logstash is restarted, ALL inputs are blocked until the one RabbitMQ input connected to the down server is connected.

I propose the following:

1. Input plugins should have their register method invoked asynchronously
1. Input plugins should have an internal state machine with the states `:starting, :running, :stopped, :permanent_error, :transient_error`. These states could be accessible via API, and state transitions would be logged.

**Comment by talevy** [comment]
by "asynchronously" you mean to have each input running on its own thread?

**Comment by andrewvc** [comment]
@talevy they currently do all run in separate threads, this means that 'register' will run in a separate thread as well. Current each input's register method is run in the main thread. If one plugin's register method is stuck, subsequent ones won't start and workers won't start either.

**Comment by talevy** [comment]
as in:

change this existing snippet[[1](https://github.com/elastic/logstash/blob/master/logstash-core/lib/logstash/pipeline.rb#L378-L386)] into something like:

```
...
@inputs.each do |input|
    input_thread = Thread.new do
        input.register
        inputworker(plugin)
    end
    @input_threads << input_thread
end
```

**Comment by andrewvc** [comment]
@talevy exactly. I do think we'd want that state machine stuff added in for debugability however.

**Comment by guyboertje** [comment]
one thing with FSMs that I find incredibly useful is storing the state history so for example (I know that the states mentioned in the original post are speculative) to know that one reached `transient error` from `running` or `starting` would be important knowledge.

**Comment by jcarterch** [comment]
@andrewvc, @talevy, I've been troubleshooting this issue and came across this and the source discussion on this problem. Is @talevy's solution usable with minor tweaking, or was this prioritized due to high level of effort? 

I'm looking to have up to 30 rabbit nodes spread out across different locations, and being unable to restart logstash without losing ingest if a single node is down is a bit scary!
